File: mobile/finance_track/lib/domain/entities/category.dart
import 'package:equatable/equatable.dart';
import 'transaction.dart';

class Category extends Equatable {
  final int? id;
  final String name;
  final String? description;
  final String? icon;
  final String? color;
  final TransactionType type;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Category({
    this.id,
    required this.name,
    this.description,
    this.icon,
    this.color,
    required this.type,
    this.isActive = true,
    required this.createdAt,
    required this.updatedAt,
  });

  @override
  List<Object?> get props => [
        id,
        name,
        description,
        icon,
        color,
        type,
        isActive,
        createdAt,
        updatedAt,
      ];

  Category copyWith({
    int? id,
    String? name,
    String? description,
    String? icon,
    String? color,
    TransactionType? type,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Category(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      icon: icon ?? this.icon,
      color: color ?? this.color,
      type: type ?? this.type,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: apps/backend/api/v1/finance/serializers/category.py
from rest_framework import serializers
from finance.models import Category

class CategorySerializer(serializers.ModelSerializer):
    """
    Serializer untuk model Category (Kategori Transaksi).
    
    Kategori digunakan untuk mengklasifikasikan transaksi keuangan menjadi
    kategori pemasukan (income) dan pengeluaran (expense).
    
    Attributes:
        name (str): Nama kategori
        type (str): Tipe kategori ('income' atau 'expense')
        icon (str): Emoji atau ikon untuk kategori (opsional)
        color (str): Kode warna hex untuk kategori (opsional)
        created_at (datetime): Waktu pembuatan (read-only)
        updated_at (datetime): Waktu update terakhir (read-only)
    """
    class Meta:
        model = Category
        fields = ['id', 'name', 'type', 'icon', 'color', 'created_at', 'updated_at']
        read_only_fields = ['created_at', 'updated_at', 'user']

    def create(self, validated_data):
        # Mengambil user dari request
        user = self.context['request'].user
        validated_data['user'] = user
        return super().create(validated_data)